pkg/controllers/jobset_controller.go

```diff
@@ -603,7 +607,174 @@ func executeSuccessPolicy(ctx context.Context, js *jobset.JobSet, ownedJobs *chi
+	if !ruleAppliesToJobFailureReason {
+		return false
+	}
+
+	parentReplicatedJob, exists := parentReplicatedJobName(failedJob)
+	if !exists {
+		// If we cannot find the parent ReplicatedJob, we assume the rule does not apply.
+		return false
+	}
+
+	ruleAppliesToParentReplicatedJob := len(rule.TargetReplicatedJobs) == 0 || slices.Contains(rule.TargetReplicatedJobs, parentReplicatedJob)
+	return ruleAppliesToParentReplicatedJob
+}
+
+// The type failurePolicyActionApplier applies a FailurePolicyAction and returns nil if the FailurePolicyAction was successfully applied.
+// The function returns an error otherwise.
+type failurePolicyActionApplier = func(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOpts *statusUpdateOpts) error
+
+// -------------------------Helper Functions for Each FailurePolicyActions Start------------------------
+
+// failJobSetActionApplier applies the FailJobSet FailurePolicyAction
+var failJobSetActionApplier failurePolicyActionApplier = func(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOpts *statusUpdateOpts) error {
+	setJobSetFailedCondition(ctx, js, constants.ReachedMaxRestartsReason, messageWithFirstFailedJob(constants.ReachedMaxRestartsMessage, matchingFailedJob.Name), updateStatusOpts)
+	return nil
+}
+
+// restartJobSetActionApplier applies the RestartJobSet FailurePolicyAction
+var restartJobSetActionApplier failurePolicyActionApplier = func(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOpts *statusUpdateOpts) error {
+	if js.Status.RestartsCountTowardsMax >= js.Spec.FailurePolicy.MaxRestarts {
+		setJobSetFailedCondition(ctx, js, constants.ReachedMaxRestartsReason, messageWithFirstFailedJob(constants.ReachedMaxRestartsMessage, matchingFailedJob.Name), updateStatusOpts)
+		return nil
+	}
+
+	shouldCountTowardsMax := true
+	failurePolicyRecreateAll(ctx, js, shouldCountTowardsMax, updateStatusOpts)
+	return nil
+}
+
+// restartJobSetAndIgnoreMaxRestartsActionApplier applies the RestartJobSetAndIgnoreMaxRestarts FailurePolicyAction
+var restartJobSetAndIgnoreMaxRestartsActionApplier failurePolicyActionApplier = func(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOpts *statusUpdateOpts) error {
+	shouldCountTowardsMax := false
+	failurePolicyRecreateAll(ctx, js, shouldCountTowardsMax, updateStatusOpts)
+	return nil
+}
+
+// -------------------------Helper Functions for Each FailurePolicyAction End--------------------------
+
+// applyFailurePolicyRuleAction applies the supplied FailurePolicyRuleAction.
+func applyFailurePolicyRuleAction(ctx context.Context, js *jobset.JobSet, matchingFailedJob *batchv1.Job, updateStatusOps *statusUpdateOpts, failurePolicyRuleAction jobset.FailurePolicyAction) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	actionFunctionMap := map[jobset.FailurePolicyAction]failurePolicyActionApplier{
```

Comment:
Once we move these to failure_policy.go let's move this to a var outside the function at the top of the file, no need to reinitialize it every time.

Reply:
This is done.